bricks/net/tcp/impl/posix.h

```diff
@@ -322,48 +323,78 @@ class ReserveLocalPortImpl final {
+    size_t save_index = index_;
+    bool keep_searching = true;
+    while (keep_searching) {
+      if (!index_) {
+        std::shuffle(std::begin(order_), std::end(order_), current::random::mt19937_64_tls());
+      }
+      const uint16_t candidate_port = [&]() {
+        uint16_t retval;
+        if (hint) {
+          retval = hint;
+          hint = 0u;
+        } else {
+          retval = order_[index_++];
+          if (index_ == order_.size()) {
+            index_ = 0u;
+          }
+          if (index_ == save_index) {
+            keep_searching = false;
+          }
+        }
+        return retval;
+      }();
+      try {
+        current::net::SocketHandle try_to_hold_port(current::net::SocketHandle::BindAndListen(), 
+                                                    BarePort(candidate_port),
+                                                    nagle_algorithm_policy,
+                                                    max_connections);
+        return current::net::ReservedLocalPort(current::net::ReservedLocalPort::Construct(),
+                                          candidate_port,
+                                          std::move(try_to_hold_port));
+      } catch (const current::net::SocketConnectException&) {
+        // Keep trying.
+        // std::cerr << "Failed in `connect`." << candidate_port << '\n';
+      } catch (const current::net::SocketBindException&) {
+        // Keep trying.
+        // std::cerr << "Failed in `bind`." << candidate_port << '\n';
+      } catch (const current::net::SocketListenException&) {
+        // Keep trying.
+        // std::cerr << "Failed in `listen`." << candidate_port << '\n';
+      }
+      // Consciously fail on other exception types here.
+    }
+    std::cerr << "FATAL ERROR: Failed to pick an available local port." << std::endl;
+    std::exit(-1);
+  }
 };
 
 }  // namespace current::net::impl
 
 // Pick an available local port.
 [[nodiscard]] inline ReservedLocalPort ReserveLocalPort(
     NagleAlgorithm nagle_algorithm_policy = kDefaultNagleAlgorithmPolicy,
-    int max_connections = kMaxServerQueuedConnections) {
-  return current::ThreadLocalSingleton<impl::ReserveLocalPortImpl>().DoIt(nagle_algorithm_policy, max_connections);
+    MaxServerQueuedConnectionsValue max_connections = kMaxServerQueuedConnections) {
+  return current::ThreadLocalSingleton<impl::ReserveLocalPortImpl>().DoIt(0, nagle_algorithm_policy, max_connections);
+}
+
+[[nodiscard]] inline ReservedLocalPort ReserveLocalPort(
+    uint16_t hint,
+    NagleAlgorithm nagle_algorithm_policy = kDefaultNagleAlgorithmPolicy,
+    MaxServerQueuedConnectionsValue max_connections = kMaxServerQueuedConnections) {
+  return current::ThreadLocalSingleton<impl::ReserveLocalPortImpl>().DoIt(hint, nagle_algorithm_policy, max_connections);
+}
+
+[[nodiscard]] inline ReservedLocalPort ReserveLocalPort(
+    NagleAlgorithm nagle_algorithm_policy,
+    uint16_t hint,
```

Comment:
Why do we need two signatures with swapped arguments?

Reply:
Yes, it's exactly the plan, for backwards compatibility (with the code that used the implementation from my dev branch that was there before).

They are all `enum class`-es, so looks safe to me.

Reply:
Compatibility with some dev branch is not a great justification for tweaking public APIs IMO

Reply:
:+1: 0280c6c6